### ISSUE TITLE
Can't update the collapse state in the references-viewlet

### ISSUE THREAD

**Issue by CsCherrYY** [microsoft/vscode#114614]
I'm developing extensions with the references-viewlet API and meet a problem. The API has only `setInput()` method to set the `SymbolTreeInput` of the reference view. 

Since a `TreeItem` could keep the collapse state (if I understand correctly), if the new `SymbolTreeInput` contains some items the same as the previous one, and the user did some changes in these items about the collapse state such as expand or collapse, the collapse state will be preserved and not apply the state from the new `SymbolTreeInput` in the new reference viewlet, as the screenshot shows.
![ref](https://user-images.githubusercontent.com/45906942/105140255-62dbda80-5b32-11eb-9d42-fe919b343494.gif)

In the screenshot, I trigger `setInput()` to set a new `SymbolTreeInput` by clicking the menu `Base on this type`. In the new `SymbolTreeInput`, The collapse state of the item `PetType` is set to `collapsed`, but it doesn't work.

It could be great if the references-viewlet API offers a method to dispose the treeView. Any other workaround would be appreciated.

**Comment by jrieken** [comment]
Do your tree items have the `id` property set? That should help but I would really need a code sample to fully understand the issue here

**Comment by CsCherrYY** [comment]
I have tried `id` and it works. Thanks! 

There is still a little confused about the `id`. I have to define a unique id although the different views could have some common items, otherwise, I will get an error telling me the item is existing.

My project is a little big, but I could introduce my code base simply. I'm developing the type hierarchy in Java by the reference viewlet. Once the user opens type hierarchy, a `SymbolTreeInput` is set by calling `setInput()`. 
```
function setTypeHierarchy(item: TypeHierarchyItem) {
    ...
    const input = new TypeHierarchyTreeInput(location, item);
    api.setInput(input);
}
```
`TypeHierarchyTreeDataProvider` will be constructed In `resolve()` of `TypeHierarchyTreeInput`, its `getTreeItem()` function will return `TreeItem` with the given `TypeHierarchyItem`.
```
export class TypeHierarchyTreeInput implements SymbolTreeInput<TypeHierarchyItem>{
    ...
    async resolve(): Promise<SymbolTreeModel<TypeHierarchyItem>> {
        ...
        const provider = new TypeHierarchyTreeDataProvider(model);
        const treeModel: SymbolTreeModel<TypeHierarchyItem> = {
		provider: provider,
		message: "",
	};
	return treeModel;
    }
}
```
```
class TypeHierarchyTreeDataProvider implements TreeDataProvider<TypeHierarchyItem> {
    ...
    async getTreeItem(element: TypeHierarchyItem): Promise<TreeItem> {
        const treeItem: TreeItem = new TreeItem(element.name);
        if (con1) {
		treeItem.collapsibleState = TreeItemCollapsibleState.None;
	} else if (con2) {
		treeItem.collapsibleState = TreeItemCollapsibleState.Expanded;
	} else {
		treeItem.collapsibleState = TreeItemCollapsibleState.Collapsed;
	}
        return treeItem;
    }
}
```
When the user clicks `Base on this type`, `setTypeHierarchy()` will be invoked with a new `TypeHierarchyItem`.

My concern is that we are overriding the old view by calling `setInput()` again. If there are some items the same as the previous view has (in this scenario, two views could have common types), I must have a proper id pattern including other information besides label. It would be better if I don't have to design and set `id` manually. Anyway, `id` works in my scenario.:smile:

**Comment by jrieken** [comment]
I understand the issue. What I am lacking (since this is just build against the TreeView-API) is a good way to reset/clear to tree. Maybe @alexr00 can help here

**Comment by alexr00** [comment]
@jrieken there isn't currently a way to reset the expand/collapse state of tree item, since we instead try hard to make sure the expand/collapse state is preserved. Sounds like a new API request...

**Comment by jrieken** [comment]
Yeah, tho request isn't just about collapse/expand but more about reset the tree. Today, this can only be achieved by using a fake content provider that returns nothing before returning the actual contents

**Comment by Eskibear** [comment]
I'm expecting VS Code to set state of tree item to `None` after it's expanded but no children returned. I don't see much sense to keep the item expandable, because currently when I re-collapse and re-expand the item, it doesn't try to fetch children again. 

Above I see @CsCherrYY was trying to re-render the item after fetching its children, assigning proper CollapseState as a workaround. Actually similar issues is also found in other TreeItem-based features. For example, call hierarchy, let's use TS to demo, proving it's a general issue across languages.

In below demo, I was expanding items one-by-one, checking if it's called by another method. It turned out only `action_1` was called by `main`. After expanding all items, I wanted to identify all leaf nodes (not called by any other method) with a glance. If you automatically remove the arrow icons (update the state to `None`) before `action_n` (n is 2,3 ... 9) , that would be very helpful. 



https://user-images.githubusercontent.com/2351748/119314969-e9065a00-bca7-11eb-8cd2-c8c8b32abdf1.mp4

**Comment by Eskibear** [comment]
If you want to keep backward compatibility and don't want to change behavior of Collpased and Expanded items. Here my proposal is to add a new state "Unknown", indicating "I don't know if it is expandable or not, unless I give it a try." 

```ts
    /**
     * Collapsible state of the tree item
     */
    export enum TreeItemCollapsibleState {
        /**
         * Determines an item can be neither collapsed nor expanded. Implies it has no children.
         */
        None = 0,
        /**
         * Determines an item is collapsed
         */
        Collapsed = 1,
        /**
         * Determines an item is expanded
         */
        Expanded = 2,
        /**
         * Determines it is unknown whether an item is expandable.
         */
        Unknown = 3
    }

```

The behavior of `Unknown` items can be:
* Render it with an right-arrow icon, indicating you can try expanding it. By clicking it, `getChildren` called.
* If one or more children are found, the case is same as `Collapsed`/`Expanded` items, and right-arrow or down-arrow icons are rendered accordingly.
* If no child is found, CollapsibleState of the item is updated to `None`, indicating it has no children. And of course the arrow icon is gone.
